Mathlib/Data/Polynomial/RingDivision.lean

```diff
@@ -394,6 +394,106 @@ theorem pow_rootMultiplicity_not_dvd {p : R[X]} (p0 : p ≠ 0) (a : R) :
     ¬(X - C a) ^ (rootMultiplicity a p + 1) ∣ p := by rw [← rootMultiplicity_le_iff p0]
 #align polynomial.pow_root_multiplicity_not_dvd Polynomial.pow_rootMultiplicity_not_dvd
 
+theorem X_sub_C_pow_dvd_iff {p : R[X]} {t : R} {n : ℕ} :
+    (X - C t) ^ n ∣ p ↔ X ^ n ∣ p.comp (X + C t) := by
```

Comment:
Glad that you completed the proof!
Actually, making use of [Polynomial.aeval](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Polynomial/AlgebraMap.html#Polynomial.aeval) may be more convenient here, as it's an AlgHom, actually an AlgEquiv in this case (composition with (X + C t) is inverse to composition with (X - C t)). Of course, divisibility is preserved by any AlgHom (even MonoidHom), which naturally yield this theorem.

Reply:
> actually an AlgEquiv in this case (composition with (X + C t) is inverse to composition with (X - C t))

In fact I have an implementation:
```lean
def algEquivAevalXAddC : R[X] ≃ₐ[R] R[X] :=
  AlgEquiv.ofAlgHom (aeval (X + C t)) (aeval (X - C t))
    (by apply algHom_ext; simp only [AlgHom.coe_comp, Function.comp_apply, aeval_X, map_sub,
      aeval_C, AlgHom.coe_id, id_eq, ← C_eq_algebraMap, add_sub_cancel])
    (by apply algHom_ext; simp only [AlgHom.coe_comp, Function.comp_apply, aeval_X, map_add,
      aeval_C, AlgHom.coe_id, id_eq, ← C_eq_algebraMap, sub_add_cancel])
```
But I can't find a way to use it to make the following proof shorter. So I haven't added it to the PR yet.

Reply:
> It may be a good idea to add MulEquivClass iff version of [map_dvd](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Divisibility/Basic.html#map_dvd).

I tried this, but it looks like that the lemmas regarding `AlgEquiv.ofAlgHom` is missing... anyway I quickly give up.

Reply:
I pushed a commit to make use of your `algEquivAevalXAddC`, as well as extracting other API. Will look at the rest later.